src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityRockBreaker.java

```diff
@@ -34,28 +39,58 @@ protected RecipeLogicEnergy createWorkable(RecipeMap<?> recipeMap) {
         return result;
     }
 
-    protected boolean checkSidesValid(BlockStaticLiquid liquid) {
-        EnumFacing frontFacing = getFrontFacing();
+    @Override
+    public void onNeighborChanged() {
+        super.onNeighborChanged();
+        if (hasValidFluids)
+            return;
+
+        boolean hasLava = false;
+        boolean hasWater = false;
         for (EnumFacing side : EnumFacing.VALUES) {
+            if (hasLava && hasWater)
+                break;
+
             if (side == frontFacing || side == EnumFacing.DOWN || side == EnumFacing.UP)
                 continue;
 
-            if (getWorld().getBlockState(getPos().offset(side)) == liquid.getDefaultState())
-                return true;
+            IBlockState state = getWorld().getBlockState(getPos().offset(side));
+            if (state == Blocks.LAVA.getDefaultState())
+                hasLava = true;
+            else if (state == Blocks.WATER.getDefaultState())
+                hasWater = true;
         }
-        return false;
+        this.hasValidFluids = hasLava && hasWater;
+    }
+
+    @Override
+    public <T> void addNotifiedInput(T input) {
+        super.addNotifiedInput(input);
+        onNeighborChanged();
+    }
+
+    @Override
+    public NBTTagCompound writeToNBT(NBTTagCompound data) {
+        super.writeToNBT(data);
+        data.setBoolean("hasValidFluids", hasValidFluids);
+        return data;
+    }
+
+    @Override
+    public void readFromNBT(NBTTagCompound data) {
+        super.readFromNBT(data);
+        this.hasValidFluids = data.getBoolean("hasValidFluids");
     }
 
-    protected static class RockBreakerRecipeLogic extends RecipeLogicEnergy {
+    protected class RockBreakerRecipeLogic extends RecipeLogicEnergy {
 
         public RockBreakerRecipeLogic(MetaTileEntity metaTileEntity, RecipeMap<?> recipeMap, Supplier<IEnergyContainer> energyContainer) {
             super(metaTileEntity, recipeMap, energyContainer);
         }
 
         @Override
         protected boolean shouldSearchForRecipes() {
-            MetaTileEntityRockBreaker rockBreaker = (MetaTileEntityRockBreaker) getMetaTileEntity();
-            return super.shouldSearchForRecipes() && rockBreaker.checkSidesValid(Blocks.LAVA) && rockBreaker.checkSidesValid(Blocks.WATER);
+            return super.shouldSearchForRecipes() && hasValidFluids;
```

Comment:
`hasValidFluids` should be checked first so it can short-circuit as `super.shouldSearchForRecipes()` is a more expensive check.